#486 - Refactoring project structure (3.3.0)

astier opened the pull request:
- Make use of StringBuilder.append() instead concatenation
- Inline variable and remove cast

reckart commented on webanno-api-annotation/src/main/java/de/tudarmstadt/ukp/clarin/webanno/api/annotation/rendering/model/VSpan.java:
```diff
@@ -40,20 +41,20 @@
             Map<String, String> aFeatures)
     {
-        this(aLayer, new VID(getAddr(aFS)), aType, asList(aOffsets), aFeatures);
+        this(aLayer, new VID(getAddr(aFS)), aType, Collections.singletonList(aOffsets), aFeatures);
```
Why is `Collections.singletonList` better than `(Collections.)asList`? asList is at least shorter ;)

astier replied:
_Arrays.asList_ with just a single argument wastes memory and is less performant than _Collections.singletonList_ because singletonList is not backed by an array and just has a reference to that one item. However, its may not worth it if you say the readability suffers disproportional. Should I revert it?

reckart replied:
I have had a look at the implementations of `asList()` and `singletonList()`. Actually, I wasn't aware that `asList()` internally simply calls `new ArrayList<>(...)`. I thought it would return a non-modifiable list. Well, always a good thing to learn. In fact, I was wishfully thinking that `asList(...)` may internally even switch between a multi-element and a single-element list implementation depending on the number of elements passed to it. I am not sure if (IntelliJ is smart enough to detect if) the switch from `asList()` to `singletonList()` safe - one is modifiable, the other is not. We could bulk-switch to`singletonList()` and fix those cases were we might get exceptions then (although I may have a hard time switching my head do that). I am a bit ambivalent... What do you think about the risk of getting exceptions due to the switch?

reckart commented on webanno-api-annotation/src/main/java/de/tudarmstadt/ukp/clarin/webanno/api/annotation/rendering/model/VSpan.java:
```diff
@@ -40,20 +41,20 @@
     {
         super(aLayer, new VID(getAddr(aFS)), aType, aEquivalenceClass, aFeatures);
-        ranges = asList(aOffsets);
+        ranges = Collections.singletonList(aOffsets);
```
See comment above.

reckart commented on webanno-api-annotation/src/main/java/de/tudarmstadt/ukp/clarin/webanno/api/annotation/rendering/model/VSpan.java:
```diff
@@ -40,20 +41,20 @@
             Map<String, String> aFeatures)
     {
-        this(aLayer, aVid, aType, asList(aOffsets), aFeatures);
+        this(aLayer, aVid, aType, Collections.singletonList(aOffsets), aFeatures);
```
See comment above.

reckart commented on webanno-brat/src/main/java/de/tudarmstadt/ukp/clarin/webanno/brat/render/model/Entity.java:
```diff
@@ -53,7 +54,7 @@ public Entity()
     public Entity(int aId, String aType, Offsets aOffsets, String aLabelText, String aColor)
     {
-        this(aId, aType, asList(aOffsets), aLabelText, aColor);
+        this(aId, aType, Collections.singletonList(aOffsets), aLabelText, aColor);
```
See comment above.

reckart commented on webanno-brat/src/main/java/de/tudarmstadt/ukp/clarin/webanno/brat/render/model/Entity.java:
```diff
@@ -63,7 +64,7 @@ public Entity(int aId, String aType, List<Offsets> aOffsets, String aLabelText,
     public Entity(VID aVid, String aType, Offsets aOffsets, String aLabelText, String aColor)
     {
-        this(aVid, aType, asList(aOffsets), aLabelText, aColor);
+        this(aVid, aType, Collections.singletonList(aOffsets), aLabelText, aColor);
```
See comment above.

reckart commented on webanno-brat/src/main/java/de/tudarmstadt/ukp/clarin/webanno/brat/render/model/EntityType.java:
```diff
@@ -62,7 +63,7 @@ public EntityType(String aName, String aLabel, String aType)
         this(aName /* name */, aType /* type */, true /* unused */, "" /* hotkey */,
                 null /* fgColor */, null /* bgColor */, null /* borderColor */,
-                Arrays.asList(aLabel) /* labels */, null /* children */, null /* attributes */, 
+            Collections.singletonList(aLabel) /* labels */, null /* children */, null /* attributes */,
```
Again, I'd go with `asList` and a static import.

reckart commented on webanno-brat/src/main/java/de/tudarmstadt/ukp/clarin/webanno/brat/render/model/RelationType.java:
```diff
@@ -66,7 +67,7 @@ public RelationType()
             String aArrowHead, String aDashArray)
     {
-        this(aColor, aArrowHead, Arrays.asList(aLabel), aType, Arrays.asList(aTarget), aDashArray);
+        this(aColor, aArrowHead, Collections.singletonList(aLabel), aType, Collections.singletonList(aTarget), aDashArray);
```
See comment above.

reckart commented on webanno-io-tsv/src/main/java/de/tudarmstadt/ukp/clarin/webanno/tsv/WebannoTsv3Reader.java:
```diff
@@ -669,13 +669,13 @@ private void createSentence(JCas aJCas, String aLine, int aBegin, int aEnd, int
 		if (aPrevEnd + 1 < aBegin) {
 		    // FIXME This is very slow. Better use StringUtils.repeat()		    
-			String pad = ""; // if there is plenty of spaces between sentences
+			StringBuilder pad = new StringBuilder(); // if there is plenty of spaces between sentences
```
Good one :)

reckart commented on webanno-ui-curation/src/main/java/de/tudarmstadt/ukp/clarin/webanno/ui/curation/component/SuggestionViewPanel.java:
```diff
@@ -618,7 +618,7 @@ private void addSuggestionColor(Project aProject, Mode aMode, Map<String, JCas>
     {
         for (ConfigurationSet cs : aCfgSet) {
-            boolean use = false;
+            boolean use;
```
I prefer having explicit initalizers for primitive variables, even if they initialize to the same value that the JVM would use.

reckart commented on webanno-io-tsv/src/main/java/de/tudarmstadt/ukp/clarin/webanno/tsv/WebannoTsv1Reader.java:
```diff
@@ -146,7 +146,7 @@ private void createDependency(JCas aJCas, Map<Integer, String> tokens,
                 // arc drawn from A to B, dependency will have (start, end) = (20, 36)
                 // arc drawn from B to A, still dependency will have (start, end) = (20, 36)
-                int begin = 0, end = 0;
+                int begin, end;
```
I prefer having explicit initalizers for primitive variables, even if they initialize to the same value that the JVM would use.

reckart commented on webanno-io-tsv/src/main/java/de/tudarmstadt/ukp/clarin/webanno/tsv/WebannoTsv1Reader.java:
```diff
@@ -101,7 +101,7 @@ private void createToken(JCas aJCas, StringBuilder text, Map<Integer, String> to
     {
         int tokenBeginPosition = 0;
-        int tokenEndPosition = 0;
+        int tokenEndPosition;
```
I prefer having explicit initalizers for primitive variables, even if they initialize to the same value that the JVM would use.

astier commented:
I adjusted the changes. I removed also the singleton stuff. I think right now the risk isn't worth the benefit.  Especially because I don't want to be responsible for weird bugs.

Besides that the PR should be ready.

reckart commented:
Thanks :)
